src/apps/config/action_codec.lua

```diff
@@ -0,0 +1,219 @@
+                       'new_link', 'link_output', 'link_input', 'stop_app',
+                       'start_app', 'reconfig_app' }
+local action_codes = {}
+for i, name in ipairs(action_names) do action_codes[name] = i end
+
+local actions = {}
+
+function actions.unlink_output (codec, appname, linkname)
+   local appname = codec:string(appname)
+   local linkname = codec:string(linkname)
+   return codec:finish(appname, linkname)
+end
+function actions.unlink_input (codec, appname, linkname)
+   local appname = codec:string(appname)
+   local linkname = codec:string(linkname)
+   return codec:finish(appname, linkname)
+end
+function actions.free_link (codec, linkspec)
+   local linkspec = codec:string(linkspec)
+   return codec:finish(linkspec)
+end
+function actions.new_link (codec, linkspec)
+   local linkspec = codec:string(linkspec)
+   return codec:finish(linkspec)
+end
+function actions.link_output (codec, appname, linkname, linkspec)
+   local appname = codec:string(appname)
+   local linkname = codec:string(linkname)
+   local linkspec = codec:string(linkspec)
+   return codec:finish(appname, linkname, linkspec)
+end
+function actions.link_input (codec, appname, linkname, linkspec)
+   local appname = codec:string(appname)
+   local linkname = codec:string(linkname)
+   local linkspec = codec:string(linkspec)
+   return codec:finish(appname, linkname, linkspec)
+end
+function actions.stop_app (codec, appname)
+   local appname = codec:string(appname)
+   return codec:finish(appname)
+end
+function actions.start_app (codec, appname, class, arg)
+   local appname = codec:string(appname)
+   local _class = codec:class(class)
+   local config = codec:config(class, arg)
+   return codec:finish(appname, _class, config)
+end
+function actions.reconfig_app (codec, appname, class, arg)
+   local appname = codec:string(appname)
+   local _class = codec:class(class)
+   local config = codec:config(class, arg)
+   return codec:finish(appname, _class, config)
+end
+
+local public_names = {}
+local function find_public_name(obj)
+   if public_names[obj] then return unpack(public_names[obj]) end
+   for modname, mod in pairs(package.loaded) do
+      if type(mod) == 'table' then
+         for name, val in pairs(mod) do
+            if val == obj then
+               if type(val) == 'table' and type(val.new) == 'function' then
+                  public_names[obj] = { modname, name }
+                  return modname, name
+               end
+            end
+         end
+      end
+   end
+   error('could not determine public name for object: '..tostring(obj))
+end
+
+local lower_case = "abcdefghijklmnopqrstuvwxyz"
+local upper_case = lower_case:upper()
+local extra = "0123456789_-"
+local alphabet = table.concat({lower_case, upper_case, extra})
+assert(#alphabet == 64)
+local function random_file_name()
+   local f = io.open('/dev/urandom', 'rb')
+   -- 22 bytes, but we only use 2^6=64 bits from each byte, so total of
```

Comment:
Calling that 64 bits is wrong, though the intention is clear; it's 6 bits. [Not a blocker, no change needed before release]